File: currency.js
// currency.js
const https = require('https');
const readline = require('readline');

const API_URL = 'https://api.exchangerate.host/latest?base=USD';
const CACHE_TTL = 60 * 1000;
let cache = { rates: null, timestamp: 0 };
let history = [];

function getRates() {
    return new Promise((resolve, reject) => {
        const now = Date.now();
        if (cache.rates && (now - cache.timestamp) < CACHE_TTL) {
            resolve(cache.rates);
            return;
        }
        https.get(API_URL, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    const json = JSON.parse(data);
                    if (json.success) {
                        cache.rates = json.rates;
                        cache.timestamp = now;
                        resolve(cache.rates);
                    } else {
                        reject(new Error('API error'));
                    }
                } catch (e) {
                    reject(e);
                }
            });
        }).on('error', reject);
    });
}

async function convert(amount, fromCur, toCur) {
    const rates = await getRates();
    if (!rates[fromCur]) throw new Error(`Currency '${fromCur}' not supported.`);
    if (!rates[toCur]) throw new Error(`Currency '${toCur}' not supported.`);
    const usdAmount = fromCur === 'USD' ? amount : amount / rates[fromCur];
    return toCur === 'USD' ? usdAmount : usdAmount * rates[toCur];
}

function addHistory(amount, fromCur, toCur, result) {
    const entry = { amount, fromCur, toCur, result, time: new Date().toLocaleTimeString() };
    history.push(entry);
    if (history.length > 10) history.shift();
}

function showHistory() {
    if (history.length === 0) {
        console.log('No conversions yet.');
        return;
    }
    console.log('\n--- History ---');
    history.forEach(h => {
        console.log(`${h.time}  ${h.amount.toFixed(2)} ${h.fromCur} = ${h.result.toFixed(2)} ${h.toCur}`);
    });
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function ask(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

async function interactive() {
    console.log('=== Currency Converter ===');
    while (true) {
        console.log('\n1. Convert');
        console.log('2. Show history');
        console.log('3. Exit');
        const choice = await ask('Choose: ');
        switch (choice.trim()) {
            case '1':
                try {
                    const amountStr = await ask('Amount: ');
                    const amount = parseFloat(amountStr);
                    if (isNaN(amount) || amount < 0) {
                        console.log('Invalid amount.');
                        continue;
                    }
                    const fromCur = (await ask('From (USD): ')).trim().toUpperCase() || 'USD';
                    const toCur = (await ask('To: ')).trim().toUpperCase();
                    if (!toCur) {
                        console.log('Please enter a target currency.');
                        continue;
                    }
                    const result = await convert(amount, fromCur, toCur);
                    console.log(`${amount.toFixed(2)} ${fromCur} = ${result.toFixed(2)} ${toCur}`);
                    addHistory(amount, fromCur, toCur, result);
                } catch (err) {
                    console.log('Error:', err.message);
                }
                break;
            case '2':
                showHistory();
                break;
            case '3':
                console.log('Goodbye!');
                rl.close();
                return;
            default:
                console.log('Invalid choice.');
        }
    }
}

async function cli() {
    const args = process.argv.slice(2);
    if (args.length !== 3) {
        console.log('Usage: node currency.js <amount> <from_currency> <to_currency>');
        process.exit(1);
    }
    const amount = parseFloat(args[0]);
    if (isNaN(amount)) {
        console.log('Invalid amount.');
        process.exit(1);
    }
    const fromCur = args[1].toUpperCase();
    const toCur = args[2].toUpperCase();
    try {
        const result = await convert(amount, fromCur, toCur);
        console.log(`${amount.toFixed(2)} ${fromCur} = ${result.toFixed(2)} ${toCur}`);
    } catch (err) {
        console.log('Error:', err.message);
        process.exit(1);
    }
}

if (process.argv.length === 2) {
    interactive();
} else {
    cli();
}

File: currency.go
// currency.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

type RatesResponse struct {
	Success bool               `json:"success"`
	Rates   map[string]float64 `json:"rates"`
}

var (
	apiURL    = "https://api.exchangerate.host/latest?base=USD"
	cache     map[string]float64
	cacheTime time.Time
	ttl       = 60 * time.Second
	history   []struct {
		Amount   float64
		From, To string
		Result   float64
		Time     string
	}
)

func getRates() (map[string]float64, error) {
	now := time.Now()
	if cache != nil && now.Sub(cacheTime) < ttl {
		return cache, nil
	}
	resp, err := http.Get(apiURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var data RatesResponse
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	if !data.Success {
		return nil, fmt.Errorf("API error")
	}
	cache = data.Rates
	cacheTime = now
	return cache, nil
}

func convert(amount float64, fromCur, toCur string) (float64, error) {
	rates, err := getRates()
	if err != nil {
		return 0, err
	}
	fromRate, ok := rates[fromCur]
	if !ok {
		return 0, fmt.Errorf("currency '%s' not supported", fromCur)
	}
	toRate, ok := rates[toCur]
	if !ok {
		return 0, fmt.Errorf("currency '%s' not supported", toCur)
	}
	usdAmount := amount
	if fromCur != "USD" {
		usdAmount = amount / fromRate
	}
	result := usdAmount
	if toCur != "USD" {
		result = usdAmount * toRate
	}
	return result, nil
}

func addHistory(amount float64, fromCur, toCur string, result float64) {
	history = append(history, struct {
		Amount   float64
		From, To string
		Result   float64
		Time     string
	}{amount, fromCur, toCur, result, time.Now().Format("15:04:05")})
	if len(history) > 10 {
		history = history[1:]
	}
}

func showHistory() {
	if len(history) == 0 {
		fmt.Println("No conversions yet.")
		return
	}
	fmt.Println("\n--- History ---")
	for _, h := range history {
		fmt.Printf("%s  %.2f %s = %.2f %s\n", h.Time, h.Amount, h.From, h.Result, h.To)
	}
}

func interactive() {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("=== Currency Converter ===")
	for {
		fmt.Println("\n1. Convert")
		fmt.Println("2. Show history")
		fmt.Println("3. Exit")
		fmt.Print("Choose: ")
		scanner.Scan()
		choice := strings.TrimSpace(scanner.Text())
		switch choice {
		case "1":
			fmt.Print("Amount: ")
			scanner.Scan()
			amountStr := strings.TrimSpace(scanner.Text())
			amount, err := strconv.ParseFloat(amountStr, 64)
			if err != nil {
				fmt.Println("Invalid amount.")
				continue
			}
			fmt.Print("From (USD): ")
			scanner.Scan()
			fromCur := strings.ToUpper(strings.TrimSpace(scanner.Text()))
			if fromCur == "" {
				fromCur = "USD"
			}
			fmt.Print("To: ")
			scanner.Scan()
			toCur := strings.ToUpper(strings.TrimSpace(scanner.Text()))
			if toCur == "" {
				fmt.Println("Please enter a target currency.")
				continue
			}
			result, err := convert(amount, fromCur, toCur)
			if err != nil {
				fmt.Println("Error:", err)
				continue
			}
			fmt.Printf("%.2f %s = %.2f %s\n", amount, fromCur, result, toCur)
			addHistory(amount, fromCur, toCur, result)
		case "2":
			showHistory()
		case "3":
			fmt.Println("Goodbye!")
			return
		default:
			fmt.Println("Invalid choice.")
		}
	}
}

func cli() {
	if len(os.Args) != 4 {
		fmt.Println("Usage: go run currency.go <amount> <from_currency> <to_currency>")
		os.Exit(1)
	}
	amount, err := strconv.ParseFloat(os.Args[1], 64)
	if err != nil {
		fmt.Println("Invalid amount.")
		os.Exit(1)
	}
	fromCur := strings.ToUpper(os.Args[2])
	toCur := strings.ToUpper(os.Args[3])
	result, err := convert(amount, fromCur, toCur)
	if err != nil {
		fmt.Println("Error:", err)
		os.Exit(1)
	}
	fmt.Printf("%.2f %s = %.2f %s\n", amount, fromCur, result, toCur)
}

func main() {
	if len(os.Args) == 1 {
		interactive()
	} else {
		cli()
	}
}
